src/rapids_singlecell/preprocessing/_qc.py

```diff
@@ -0,0 +1,434 @@
+            qc_vars = [qc_vars]
+        for qc_var in qc_vars:
+            mask = cp.array(adata.var[qc_var], dtype=cp.bool_)
+            sums_cells_sub = _second_pass_qc(X, mask)
+
+            adata.obs[f"total_{expr_type}_{qc_var}"] = cp.asnumpy(sums_cells_sub)
+            adata.obs[f"pct_{expr_type}_{qc_var}"] = cp.asnumpy(
+                sums_cells_sub / sums_cells * 100
+            )
+            if log1p:
+                adata.obs[f"log1p_total_{expr_type}_{qc_var}"] = cp.asnumpy(
+                    cp.log1p(sums_cells_sub)
+                )
+
+
+def _first_pass_qc(X):
+    if isinstance(X, DaskArray):
+        return _first_pass_qc_dask(X)
+
+    sums_cells = cp.zeros(X.shape[0], dtype=X.dtype)
+    sums_genes = cp.zeros(X.shape[1], dtype=X.dtype)
+    cell_ex = cp.zeros(X.shape[0], dtype=cp.int32)
+    gene_ex = cp.zeros(X.shape[1], dtype=cp.int32)
+    if sparse.issparse(X):
+        if sparse.isspmatrix_csr(X):
+            from ._kernels._qc_kernels import _sparse_qc_csr
+
+            block = (32,)
+            grid = (int(math.ceil(X.shape[0] / block[0])),)
+            sparse_qc_csr = _sparse_qc_csr(X.data.dtype)
+            sparse_qc_csr(
+                grid,
+                block,
+                (
+                    X.indptr,
+                    X.indices,
+                    X.data,
+                    sums_cells,
+                    sums_genes,
+                    cell_ex,
+                    gene_ex,
+                    X.shape[0],
+                ),
+            )
+        elif sparse.isspmatrix_csc(X):
+            from ._kernels._qc_kernels import _sparse_qc_csc
+
+            block = (32,)
+            grid = (int(math.ceil(X.shape[1] / block[0])),)
+            sparse_qc_csc = _sparse_qc_csc(X.data.dtype)
+            sparse_qc_csc(
+                grid,
+                block,
+                (
+                    X.indptr,
+                    X.indices,
+                    X.data,
+                    sums_cells,
+                    sums_genes,
+                    cell_ex,
+                    gene_ex,
+                    X.shape[1],
+                ),
+            )
+        else:
+            raise ValueError("Please use a csr or csc matrix")
+    else:
+        from ._kernels._qc_kernels import _sparse_qc_dense
+
+        if not X.flags.c_contiguous:
+            X = cp.asarray(X, order="C")
+        block = (16, 16)
+        grid = (
+            int(math.ceil(X.shape[0] / block[0])),
+            int(math.ceil(X.shape[1] / block[1])),
+        )
+        sparse_qc_dense = _sparse_qc_dense(X.dtype)
+        sparse_qc_dense(
+            grid,
+            block,
+            (X, sums_cells, sums_genes, cell_ex, gene_ex, X.shape[0], X.shape[1]),
+        )
+    return sums_cells, sums_genes, cell_ex, gene_ex
+
+
+@with_cupy_rmm
+def _first_pass_qc_dask(X):
+    import dask
+    import dask.array as da
+
+    if isinstance(X._meta, sparse.csr_matrix):
```

Comment:
I can’t visually parse this. To me this looks like 4 times the same code block repeated across 2 if branches with some subtle differences.

Should be refactored so the differences between the 4 blocks provide the majority of the signal and don‘t get lost in the noise.

Reply:
Since we can't return more than 1 array we had to split obs from var calc. In addition to that we had to split dense from sparse.

Reply:
please figure out a way to reuse code here.